crates/kornia-pnp/src/camera.rs

```diff
@@ -0,0 +1,392 @@
+        // Check that the matrix has the expected form
+        if k[0][1] != 0.0 || k[1][0] != 0.0 || k[2][0] != 0.0 || k[2][1] != 0.0 || k[2][2] != 1.0 {
+            return Err(CameraError::InvalidIntrinsics(
+                "Intrinsics matrix must have form [[fx, 0, cx], [0, fy, cy], [0, 0, 1]]"
+                    .to_string(),
+            ));
+        }
+
+        Ok(Self {
+            fx: k[0][0],
+            fy: k[1][1],
+            cx: k[0][2],
+            cy: k[1][2],
+        })
+    }
+
+    /// Convert to 3x3 intrinsics matrix.
+    pub fn to_matrix(&self) -> [[f32; 3]; 3] {
+        [
+            [self.fx, 0.0, self.cx],
+            [0.0, self.fy, self.cy],
+            [0.0, 0.0, 1.0],
+        ]
+    }
+}
+
+/// Represents polynomial distortion parameters using the Brown-Conrady model.
+#[derive(Debug, Clone, PartialEq)]
+#[allow(missing_docs)]
+pub struct PolynomialDistortion {
+    /// Radial distortion coefficients
+    pub k1: f32,
+    pub k2: f32,
+    pub k3: f32,
+    pub k4: f32,
+    pub k5: f32,
+    pub k6: f32,
+    /// Tangential distortion coefficients
+    pub p1: f32,
+    pub p2: f32,
+}
+
+impl PolynomialDistortion {
+    /// Create distortion parameters with all coefficients set to zero (no distortion).
+    pub fn none() -> Self {
+        Self {
+            k1: 0.0,
+            k2: 0.0,
+            k3: 0.0,
+            k4: 0.0,
+            k5: 0.0,
+            k6: 0.0,
+            p1: 0.0,
+            p2: 0.0,
+        }
+    }
+
+    /// Create distortion parameters with only first two radial coefficients.
+    pub fn radial(k1: f32, k2: f32) -> Self {
+        Self {
+            k1,
+            k2,
+            k3: 0.0,
+            k4: 0.0,
+            k5: 0.0,
+            k6: 0.0,
+            p1: 0.0,
+            p2: 0.0,
+        }
+    }
+
+    /// Create distortion parameters with radial and tangential coefficients.
+    pub fn radial_tangential(k1: f32, k2: f32, p1: f32, p2: f32) -> Self {
+        Self {
+            k1,
+            k2,
+            k3: 0.0,
+            k4: 0.0,
+            k5: 0.0,
+            k6: 0.0,
+            p1,
+            p2,
+        }
+    }
+
+    /// Check if there is any distortion.
+    pub fn has_distortion(&self) -> bool {
+        self.k1 != 0.0
+            || self.k2 != 0.0
+            || self.k3 != 0.0
+            || self.k4 != 0.0
+            || self.k5 != 0.0
+            || self.k6 != 0.0
+            || self.p1 != 0.0
+            || self.p2 != 0.0
+    }
+}
+
+/// A complete camera model with intrinsics and optional distortion.
+#[derive(Debug, Clone)]
+pub struct CameraModel {
+    /// Camera intrinsics
+    pub intrinsics: CameraIntrinsics,
+    /// Distortion parameters (None for no distortion)
+    pub distortion: Option<PolynomialDistortion>,
+}
+
+impl CameraModel {
+    /// Create a camera model without distortion.
+    pub fn pinhole(intrinsics: CameraIntrinsics) -> Self {
```

Comment:
wondering if the type should be known at compile time instead to use full rust type safe checks semantics /cc @cjpurackal